lib/types.ts

```diff
@@ -1,7 +1,23 @@
 import type { JSONSchema4 } from "json-schema";
 import type { Rule } from "eslint";
-export interface RuleListener {
-  [key: string]: ((node: never) => void) | undefined;
+import type { AST } from "jsonc-eslint-parser";
+
+export type RuleFunction<Node extends AST.JSONNode = never> = (
+  node: Node,
+) => void;
+
+export type BuiltInRuleListeners = {
+  [Node in AST.JSONNode as Node["type"]]?: RuleFunction<Node>;
+};
+
+export type BuiltInRuleListenerExits = {
+  [Node in AST.JSONNode as `${Node["type"]}:exit`]?: RuleFunction<Node>;
+};
+
+export interface RuleListener
```

Comment:
Is this considered a public API? If so, we'll need a changesets entry, right?

Reply:
Does eslint-plugin-package-json also use the type annotations defined here? If so, I think it would be a good idea to export it from index.ts and mark it as `feat`.

Reply:
It doesn't, but I'd love to reuse them in some way. eslint-plugin-package-json doesn't have a dependency on eslint-plugin-jsonc though. Would it make sense to push these into jsonc-eslint-parser as a type utility?

Reply:
> Would it make sense to push these into jsonc-eslint-parser as an exported type utility?

I like that idea!